File: plasma-reaper/app/utils/protocol.js
/**
 * protocol.js — PlasmaReaper binary protocol implementation
 * Author: jayis1
 * License: MIT
 *
 * Implements the framing, CRC-8, and serialization for the
 * PlasmaReaper USB CDC / BLE command protocol.
 */

// Command codes
export const CMD_PING = 0x0D;
export const CMD_GET_VERSION = 0x0E;
export const CMD_FIRE = 0x01;
export const CMD_SET_PARAMS = 0x02;
export const CMD_GET_RESULT = 0x03;
export const CMD_START_SWEEP = 0x04;
export const CMD_STOP_SWEEP = 0x05;
export const CMD_GET_WAVEFORM = 0x06;
export const CMD_SET_TRIGGER = 0x07;
export const CMD_SET_PATTERNS = 0x08;
export const CMD_GET_STATUS = 0x09;

// Response codes
export const RESP_OK = 0x80;
export const RESP_ERROR = 0x81;
export const RESP_RESULT = 0x82;
export const RESP_SWEEP_PROGRESS = 0x83;
export const RESP_SWEEP_COMPLETE = 0x84;
export const RESP_STATUS = 0x85;
export const RESP_WAVEFORM = 0x86;
export const RESP_VERSION = 0x87;

// Frame delimiters
const FRAME_START = 0xA5;
const FRAME_END = 0x5A;

/**
 * Compute CRC-8 (polynomial 0x07, init 0x00)
 */
function crc8(data) {
  let crc = 0x00;
  for (let i = 0; i < data.length; i++) {
    crc ^= data[i];
    for (let j = 0; j < 8; j++) {
      if (crc & 0x80) {
        crc = ((crc << 1) ^ 0x07) & 0xFF;
      } else {
        crc = (crc << 1) & 0xFF;
      }
    }
  }
  return crc;
}

/**
 * Build a command frame.
 * @param {number} cmd - Command code
 * @param {Uint8Array|null} payload - Payload bytes
 * @returns {Uint8Array} Complete frame with delimiters and CRC
 */
export function buildFrame(cmd, payload) {
  const plen = payload ? payload.length : 0;
  const frame = new Uint8Array(5 + plen + 1);

  frame[0] = FRAME_START;
  frame[1] = cmd;
  frame[2] = plen & 0xFF;
  frame[3] = (plen >> 8) & 0xFF;

  if (payload) {
    frame.set(payload, 4);
  }

  // CRC over cmd + len + payload
  const crcData = new Uint8Array(3 + plen);
  crcData[0] = cmd;
  crcData[1] = plen & 0xFF;
  crcData[2] = (plen >> 8) & 0xFF;
  if (payload) {
    crcData.set(payload, 3);
  }
  frame[4 + plen] = crc8(crcData);
  frame[5 + plen] = FRAME_END;

  return frame;
}

/**
 * Parse a response frame.
 * @param {Uint8Array} raw - Raw bytes received from device
 * @returns {object|null} { type, data } or null if invalid
 */
export function parseFrame(raw) {
  if (!raw || raw.length < 7) return null;
  if (raw[0] !== FRAME_START) return null;
  if (raw[raw.length - 1] !== FRAME_END) return null;

  const cmd = raw[1];
  const plen = raw[2] | (raw[3] << 8);
  const payload = raw.slice(4, 4 + plen);
  const crc = raw[4 + plen];

  // Verify CRC
  const crcData = new Uint8Array(3 + plen);
  crcData[0] = cmd;
  crcData[1] = plen & 0xFF;
  crcData[2] = (plen >> 8) & 0xFF;
  crcData.set(payload, 3);

  if (crc8(crcData) !== crc) {
    console.error('CRC mismatch in response frame');
    return null;
  }

  // Parse based on response type
  switch (cmd) {
  case RESP_OK:
    return { type: RESP_OK, data: { success: payload[0] === 1 } };

  case RESP_RESULT:
    return {
      type: RESP_RESULT,
      data: {
        outcome: payload[0],
        shotsFired: payload[1] | (payload[2] << 8) | (payload[3] << 16) | (payload[4] << 24),
        elapsedUs: payload[5] | (payload[6] << 8) | (payload[7] << 16) | (payload[8] << 24),
        waveformLen: payload[9] | (payload[10] << 8),
        targetResponse: String.fromCharCode.apply(null, payload.slice(11)),
      },
    };

  case RESP_STATUS:
    return {
      type: RESP_STATUS,
      data: {
        sweepRunning: payload[0] === 1,
        totalCells: payload[1] | (payload[2] << 8) | (payload[3] << 16) | (payload[4] << 24),
        completedCells: payload[5] | (payload[6] << 8) | (payload[7] << 16) | (payload[8] << 24),
        successCount: payload[9] | (payload[10] << 8) | (payload[11] << 16) | (payload[12] << 24),
        totalShots: payload[13] | (payload[14] << 8) | (payload[15] << 16) | (payload[16] << 24),
        successShots: payload[17] | (payload[18] << 8) | (payload[19] << 16) | (payload[20] << 24),
      },
    };

  case RESP_SWEEP_PROGRESS:
    return {
      type: RESP_SWEEP_PROGRESS,
      data: {
        completedCells: payload[0] | (payload[1] << 8) | (payload[2] << 16) | (payload[3] << 24),
        successCount: payload[4] | (payload[5] << 8) | (payload[6] << 16) | (payload[7] << 24),
        outcome: payload[8],
        targetResponse: String.fromCharCode.apply(null, payload.slice(9)),
      },
    };

  case RESP_SWEEP_COMPLETE:
    return {
      type: RESP_SWEEP_COMPLETE,
      data: {
        totalCells: payload[0] | (payload[1] << 8) | (payload[2] << 16) | (payload[3] << 24),
        successCount: payload[4] | (payload[5] << 8) | (payload[6] << 16) | (payload[7] << 24),
        failureCount: payload[8] | (payload[9] << 8) | (payload[10] << 16) | (payload[11] << 24),
      },
    };

  case RESP_WAVEFORM:
    return {
      type: RESP_WAVEFORM,
      data: {
        len: payload[0] | (payload[1] << 8),
        samples: parseWaveformSamples(payload.slice(2)),
      },
    };

  case RESP_VERSION:
    return {
      type: RESP_VERSION,
      data: {
        major: payload[0],
        minor: payload[1],
        patch: payload[2],
      },
    };

  default:
    return { type: cmd, data: payload };
  }
}

function parseWaveformSamples(bytes) {
  const samples = [];
  for (let i = 0; i + 1 < bytes.length; i += 2) {
    samples.push(bytes[i] | (bytes[i + 1] << 8));
  }
  return samples;
}

/**
 * Read a BLE response (monitor the notification characteristic).
 */
export async function readBleResponse(device, serviceUuid, charUuid) {
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => reject(new Error('BLE read timeout')), 5000);

    device.monitorCharacteristicForService(
      serviceUuid, charUuid,
      (error, characteristic) => {
        if (error) {
          clearTimeout(timeout);
          reject(error);
          return;
        }
        if (characteristic && characteristic.value) {
          clearTimeout(timeout);
          const raw = base64ToUint8Array(characteristic.value);
          resolve(raw);
        }
      }
    );
  });
}

function base64ToUint8Array(base64) {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
  const lookup = new Array(256);
  for (let i = 0; i < chars.length; i++) lookup[chars.charCodeAt(i)] = i;

  const bytes = [];
  for (let i = 0; i < base64.length; i += 4) {
    const a = lookup[base64.charCodeAt(i)] || 0;
    const b = lookup[base64.charCodeAt(i + 1)] || 0;
    const c = lookup[base64.charCodeAt(i + 2)] || 0;
    const d = lookup[base64.charCodeAt(i + 3)] || 0;
    bytes.push((a << 2) | (b >> 4));
    if (base64[i + 2] !== '=') bytes.push(((b & 0xF) << 4) | (c >> 2));
    if (base64[i + 3] !== '=') bytes.push(((c & 0x3) << 6) | d);
  }
  return new Uint8Array(bytes);
}

/**
 * Serialize glitch parameters to a Uint8Array payload.
 */
export function serializeParams(params) {
  const p = new Uint8Array(31);
  p[0] = params.vectorMask;
  writeU32(p, 1, params.triggerOffsetNs);
  writeU32(p, 5, params.glitchWidthNs);
  writeU16(p, 9, params.powerDepthMv);
  p[11] = params.powerSeriesR;
  p[12] = params.clockShape;
  writeU32(p, 13, params.clockCycleOffset);
  writeU16(p, 17, params.emPulseMv);
  writeU16(p, 19, params.emPulseWidthNs);
  writeU32(p, 21, params.interVectorNs);
  writeU16(p, 25, params.repeatCount);
  writeU32(p, 27, params.repeatDelayNs);
  return p;
}

function writeU32(arr, offset, val) {
  arr[offset] = val & 0xFF;
  arr[offset + 1] = (val >> 8) & 0xFF;
  arr[offset + 2] = (val >> 16) & 0xFF;
  arr[offset + 3] = (val >> 24) & 0xFF;
}

function writeU16(arr, offset, val) {
  arr[offset] = val & 0xFF;
  arr[offset + 1] = (val >> 8) & 0xFF;
}

/**
 * Outcome labels for display.
 */
export const OUTCOME_LABELS = {
  0: 'Pending',
  1: 'SUCCESS',
  2: 'Failure',
  3: 'No Response',
  4: 'Glitch Invalid',
  5: 'Timeout',
};

export const OUTCOME_COLORS = {
  0: '#999',
  1: '#4CAF50',
  2: '#F44336',
  3: '#FF9800',
  4: '#9C27B0',
  5: '#607D8B',
};

File: plasma-reaper/firmware/drivers/protocol.c
/*
 * protocol.c — Host command protocol parser
 * PlasmaReaper Multi-Vector Fault Injection Toolkit
 *
 * Author: jayis1
 * License: GPL-2.0
 *
 * Parses incoming command frames from the host (USB CDC or BLE) and
 * dispatches them to the appropriate handlers. Frames use a simple
 * binary format:
 *
 *   [0xA5] [cmd] [len_lo] [len_hi] [payload...] [crc8] [0x5A]
 *
 * Responses use the same format with the response code in the cmd
 * field.
 */

#include "protocol.h"
#include "usb_cdc.h"
#include "ble_if.h"
#include "../registers.h"
#include "../board.h"

/* ---- Private state ------------------------------------------------- */

#define MAX_PAYLOAD 256

static uint8_t  g_rx_frame[260];
static uint16_t g_rx_idx;
static bool     g_in_frame;
static uint16_t g_payload_len;

/* ---- CRC-8 (poly 0x07, init 0x00) ---------------------------------- */

static uint8_t crc8(const uint8_t *data, uint16_t len)
{
    uint8_t crc = 0x00;
    for (uint16_t i = 0; i < len; i++) {
        crc ^= data[i];
        for (int j = 0; j < 8; j++) {
            if (crc & 0x80)
                crc = (crc << 1) ^ 0x07;
            else
                crc <<= 1;
        }
    }
    return crc;
}

/* ---- Frame transmission -------------------------------------------- */

static void send_frame(uint8_t code, const uint8_t *payload, uint16_t len)
{
    uint8_t frame[260];
    uint16_t idx = 0;

    frame[idx++] = FRAME_START_BYTE;
    frame[idx++] = code;
    frame[idx++] = len & 0xFF;
    frame[idx++] = (len >> 8) & 0xFF;
    for (uint16_t i = 0; i < len; i++)
        frame[idx++] = payload[i];
    frame[idx++] = crc8(&frame[1], 3 + len); /* CRC over cmd+len+payload */
    frame[idx++] = FRAME_END_BYTE;

    /* Send over USB if connected */
    if (usb_cdc_connected())
        usb_cdc_tx(frame, idx);
    /* Also send over BLE if connected */
    if (ble_if_connected())
        ble_if_tx(frame, idx);
}

/* ---- Payload deserialization helpers ------------------------------- */

static void deserialize_params(const uint8_t *p, glitch_params_t *params)
{
    params->vector_mask       = p[0];
    params->trigger_offset_ns = (uint32_t)p[1] | ((uint32_t)p[2] << 8)
                              | ((uint32_t)p[3] << 16) | ((uint32_t)p[4] << 24);
    params->glitch_width_ns   = (uint32_t)p[5] | ((uint32_t)p[6] << 8)
                              | ((uint32_t)p[7] << 16) | ((uint32_t)p[8] << 24);
    params->power_depth_mv    = (uint16_t)p[9] | ((uint16_t)p[10] << 8);
    params->power_series_r    = p[11];
    params->clock_shape       = p[12];
    params->clock_cycle_offset = (uint32_t)p[13] | ((uint32_t)p[14] << 8)
                              | ((uint32_t)p[15] << 16) | ((uint32_t)p[16] << 24);
    params->em_pulse_mv       = (uint16_t)p[17] | ((uint16_t)p[18] << 8);
    params->em_pulse_width_ns = (uint16_t)p[19] | ((uint16_t)p[20] << 8);
    params->inter_vector_ns   = (uint32_t)p[21] | ((uint32_t)p[22] << 8)
                              | ((uint32_t)p[23] << 16) | ((uint32_t)p[24] << 24);
    params->repeat_count      = (uint16_t)p[25] | ((uint16_t)p[26] << 8);
    params->repeat_delay_ns   = (uint32_t)p[27] | ((uint32_t)p[28] << 8)
                              | ((uint32_t)p[29] << 16) | ((uint32_t)p[30] << 24);
}

static void serialize_result(const glitch_result_t *result, uint8_t *p)
{
    p[0] = (uint8_t)result->outcome;
    p[1] = result->shots_fired & 0xFF;
    p[2] = (result->shots_fired >> 8) & 0xFF;
    p[3] = (result->shots_fired >> 16) & 0xFF;
    p[4] = (result->shots_fired >> 24) & 0xFF;
    p[5] = result->elapsed_us & 0xFF;
    p[6] = (result->elapsed_us >> 8) & 0xFF;
    p[7] = (result->elapsed_us >> 16) & 0xFF;
    p[8] = (result->elapsed_us >> 24) & 0xFF;
    p[9] = result->waveform_len & 0xFF;
    p[10] = (result->waveform_len >> 8) & 0xFF;
    /* Target response: up to 128 bytes, null-terminated */
    for (int i = 0; i < 128 && result->target_response[i]; i++)
        p[11 + i] = (uint8_t)result->target_response[i];
}

/* ---- Command handlers ---------------------------------------------- */

/* These are implemented in main.c */
void protocol_handle_fire(glitch_params_t *params, glitch_result_t *result);
void protocol_handle_start_sweep(sweep_config_t *config);
void protocol_handle_stop_sweep(void);
void protocol_handle_set_trigger(trigger_config_t *config);
void protocol_handle_set_patterns(pattern_config_t *patterns);
void protocol_handle_get_status(sweep_status_t *status,
                                glitch_result_t *last_result,
                                uint32_t *total_shots,
                                uint32_t *success_shots);
void protocol_handle_get_waveform(uint16_t *waveform, uint16_t *len);

/* ---- Process a complete frame -------------------------------------- */

static void process_frame(uint8_t cmd, const uint8_t *payload, uint16_t len)
{
    switch (cmd) {
    case CMD_PING: {
        uint8_t resp = 0x01;
        send_frame(RESP_OK, &resp, 1);
        break;
    }

    case CMD_GET_VERSION: {
        uint8_t version[4] = { 1, 0, 0, 0 }; /* v1.0.0 */
        send_frame(RESP_VERSION, version, 4);
        break;
    }

    case CMD_FIRE: {
        if (len < 31)
            break;
        glitch_params_t params;
        glitch_result_t result;
        deserialize_params(payload, &params);
        protocol_handle_fire(&params, &result);
        uint8_t resp[139];
        serialize_result(&result, resp);
        send_frame(RESP_RESULT, resp, 139);
        break;
    }

    case CMD_START_SWEEP: {
        if (len < sizeof(sweep_config_t))
            break;
        sweep_config_t config;
        /* Copy raw bytes (assumes same layout on host and device) */
        uint8_t *dst = (uint8_t *)&config;
        for (uint16_t i = 0; i < sizeof(config) && i < len; i++)
            dst[i] = payload[i];
        protocol_handle_start_sweep(&config);
        uint8_t resp = 0x01;
        send_frame(RESP_OK, &resp, 1);
        break;
    }

    case CMD_STOP_SWEEP:
        protocol_handle_stop_sweep();
        uint8_t resp = 0x01;
        send_frame(RESP_OK, &resp, 1);
        break;

    case CMD_GET_STATUS: {
        sweep_status_t status;
        glitch_result_t last_result;
        uint32_t total, success;
        protocol_handle_get_status(&status, &last_result, &total, &success);
        uint8_t resp[32];
        resp[0] = status.running ? 1 : 0;
        resp[1] = status.total_cells & 0xFF;
        resp[2] = (status.total_cells >> 8) & 0xFF;
        resp[3] = (status.total_cells >> 16) & 0xFF;
        resp[4] = (status.total_cells >> 24) & 0xFF;
        resp[5] = status.completed_cells & 0xFF;
        resp[6] = (status.completed_cells >> 8) & 0xFF;
        resp[7] = (status.completed_cells >> 16) & 0xFF;
        resp[8] = (status.completed_cells >> 24) & 0xFF;
        resp[9] = status.success_count & 0xFF;
        resp[10] = (status.success_count >> 8) & 0xFF;
        resp[11] = (status.success_count >> 16) & 0xFF;
        resp[12] = (status.success_count >> 24) & 0xFF;
        resp[13] = total & 0xFF;
        resp[14] = (total >> 8) & 0xFF;
        resp[15] = (total >> 16) & 0xFF;
        resp[16] = (total >> 24) & 0xFF;
        resp[17] = success & 0xFF;
        resp[18] = (success >> 8) & 0xFF;
        resp[19] = (success >> 16) & 0xFF;
        resp[20] = (success >> 24) & 0xFF;
        send_frame(RESP_STATUS, resp, 21);
        break;
    }

    case CMD_GET_WAVEFORM: {
        uint16_t waveform[256];
        uint16_t wlen;
        protocol_handle_get_waveform(waveform, &wlen);
        uint8_t resp[514];
        resp[0] = wlen & 0xFF;
        resp[1] = (wlen >> 8) & 0xFF;
        for (uint16_t i = 0; i < wlen; i++) {
            resp[2 + i * 2] = waveform[i] & 0xFF;
            resp[2 + i * 2 + 1] = (waveform[i] >> 8) & 0xFF;
        }
        send_frame(RESP_WAVEFORM, resp, 2 + wlen * 2);
        break;
    }

    case CMD_SET_TRIGGER: {
        trigger_config_t config;
        uint8_t *dst = (uint8_t *)&config;
        for (uint16_t i = 0; i < sizeof(config) && i < len; i++)
            dst[i] = payload[i];
        protocol_handle_set_trigger(&config);
        uint8_t resp = 0x01;
        send_frame(RESP_OK, &resp, 1);
        break;
    }

    default:
        /* Unknown command */
        break;
    }
}

/* ---- Frame parser (state machine) ---------------------------------- */

static void process_byte(uint8_t byte)
{
    if (!g_in_frame) {
        if (byte == FRAME_START_BYTE) {
            g_in_frame = true;
            g_rx_idx = 0;
            g_rx_frame[g_rx_idx++] = byte;
        }
        return;
    }

    if (g_rx_idx >= sizeof(g_rx_frame)) {
        /* Overflow — reset */
        g_in_frame = false;
        return;
    }

    g_rx_frame[g_rx_idx++] = byte;

    /* After we have start + cmd + len (4 bytes), we know the payload length */
    if (g_rx_idx == 4) {
        g_payload_len = g_rx_frame[2] | (g_rx_frame[3] << 8);
        if (g_payload_len > MAX_PAYLOAD) {
            g_in_frame = false;
            return;
        }
    }

    /* Check for end of frame:
     * total = 1 (start) + 1 (cmd) + 2 (len) + payload + 1 (crc) + 1 (end) */
    if (g_rx_idx == 5 + g_payload_len + 1) {
        /* This should be the CRC byte */
        /* Next byte should be end byte */
    }

    if (g_rx_idx == 5 + g_payload_len + 2) {
        /* Full frame received */
        if (byte == FRAME_END_BYTE) {
            /* Verify CRC */
            uint8_t expected_crc = crc8(&g_rx_frame[1], 3 + g_payload_len);
            uint8_t received_crc = g_rx_frame[5 + g_payload_len - 1 + 1 - 1];
            if (expected_crc == received_crc) {
                /* Valid frame — process it */
                process_frame(g_rx_frame[1],
                              &g_rx_frame[4],
                              g_payload_len);
            }
        }
        g_in_frame = false;
    }
}

/* ---- Poll (called from main loop) ---------------------------------- */

void protocol_poll(void)
{
    uint8_t buf[64];
    uint32_t count;

    /* Read from USB */
    count = usb_cdc_rx(buf, sizeof(buf));
    for (uint32_t i = 0; i < count; i++)
        process_byte(buf[i]);

    /* Read from BLE */
    count = ble_if_rx(buf, sizeof(buf));
    for (uint32_t i = 0; i < count; i++)
        process_byte(buf[i]);
}

/* ---- Sweep progress reporting -------------------------------------- */

void protocol_send_sweep_progress(const sweep_status_t *status,
                                  const glitch_result_t *result)
{
    uint8_t payload[160];
    uint16_t idx = 0;

    /* Sweep status */
    payload[idx++] = status->completed_cells & 0xFF;
    payload[idx++] = (status->completed_cells >> 8) & 0xFF;
    payload[idx++] = (status->completed_cells >> 16) & 0xFF;
    payload[idx++] = (status->completed_cells >> 24) & 0xFF;
    payload[idx++] = status->success_count & 0xFF;
    payload[idx++] = (status->success_count >> 8) & 0xFF;
    payload[idx++] = (status->success_count >> 16) & 0xFF;
    payload[idx++] = (status->success_count >> 24) & 0xFF;

    /* Last result */
    payload[idx++] = (uint8_t)result->outcome;

    /* Target response */
    for (int i = 0; i < 128 && result->target_response[i]; i++)
        payload[idx++] = (uint8_t)result->target_response[i];

    send_frame(RESP_SWEEP_PROGRESS, payload, idx);
}

void protocol_send_sweep_complete(const sweep_status_t *status)
{
    uint8_t payload[20];
    payload[0] = status->total_cells & 0xFF;
    payload[1] = (status->total_cells >> 8) & 0xFF;
    payload[2] = (status->total_cells >> 16) & 0xFF;
    payload[3] = (status->total_cells >> 24) & 0xFF;
    payload[4] = status->success_count & 0xFF;
    payload[5] = (status->success_count >> 8) & 0xFF;
    payload[6] = (status->success_count >> 16) & 0xFF;
    payload[7] = (status->success_count >> 24) & 0xFF;
    payload[8] = status->failure_count & 0xFF;
    payload[9] = (status->failure_count >> 8) & 0xFF;
    payload[10] = (status->failure_count >> 16) & 0xFF;
    payload[11] = (status->failure_count >> 24) & 0xFF;

    send_frame(RESP_SWEEP_COMPLETE, payload, 12);
}

/* ---- End of file --------------------------------------------------- */